---
title: "split_markers.Rmd"
author: "Eily Allan"
date: "7/29/2021"
output: html_document
params:
  run.num:
    value: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(filesstrings)
library(here)

fastq_noprimers_folder <- paste0(here("Output","cutadapt_output"),"/run",params$run.num,"_",format(Sys.Date(), "%Y%m%d"), "/noprimers")

```

## Split primer-trimmed fastq files by marker 

In between cutadapt and dada2, need to split sequencing runs by marker so dada2 is only using one marker at a time. That way can customize the amount to trim in dada2 by marker. 

```{r find all files and split by marker}

# use the pattern to sort out the files - ^ means "starts with" so we can use that to separate by marker
COI_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_COI", full.names = T)
MF_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_MiFish", full.names = T)
MM_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_MiMammal", full.names = T)

# then we actually want to create subfolders for each marker and put move the files into each folder by marker
new_COI_folder <- paste0(fastq_noprimers_folder,"/COI")
new_MF_folder <- paste0(fastq_noprimers_folder,"/MiFish")
new_MM_folder <- paste0(fastq_noprimers_folder,"/MiMammal")

# then actually create them
dir.create(new_COI_folder)
dir.create(new_MF_folder)
dir.create(new_MM_folder)

# then we  need to move the files (which we already have listed out)
file.move(COI_files, new_COI_folder)
file.move(MF_files, new_MF_folder)
file.move(MM_files, new_MM_folder)

```

## Split metadata files 

Now we should also split the sequencing metadata file by marker to use as input into dada2

```{r split metadata file by marker}
cutadapt_output_metadata_path <- paste0(fastq_noprimers_folder, "/output.metadata.csv")
cutadapt_output_metadata_all <- read.csv(cutadapt_output_metadata_path)

# split up based on marker
cutadapt_output_metadata_COI <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "COI",]
cutadapt_output_metadata_MF <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "MiFish",]
cutadapt_output_metadata_MM <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "MiMammal",]

# write them all to csv files 
write.csv(cutadapt_output_metadata_COI, paste0(new_COI_folder, "/cutadapt_output_metadata_COI.csv"), row.names=FALSE)
write.csv(cutadapt_output_metadata_MF, paste0(new_MF_folder, "/cutadapt_output_metadata_MiFish.csv"), row.names=FALSE)
write.csv(cutadapt_output_metadata_MM, paste0(new_MM_folder, "/cutadapt_output_metadata_MiMammal.csv"), row.names=FALSE)


```

