---
title: "split_markers.Rmd"
author: "Eily Allan"
date: "2/22/22"
output: html_document
params:
  run.num:
    value: 6
---

Last updated: 2/2/22

# Overview

Cutadapt can be run with several different primer sets because we tell the software what primers to cut off each sample. After primers have been removed by cutadapt, the next step is to generate ASVs using dada2. However, dada2 should be run with only one primer at a time. One reason to do this is so that we can give different trimming lengths becuase we have different sized fragments with the different markers. Also because it is comparing sequences and making ASVs, we should probably split them up beforehand even for similar 12S MiFish and MiMammal so we don't mess up the alogrithm. (It will also run a lot faster if you split them up by marker.) So this script takes the cutadapt output and splits it up into subfolders for each primer that we can then read in to our dada2 scripts marker by marker. 

# Set up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(filesstrings)
library(here)

fastq_noprimers_folder <- paste0("/Users/elizabethandruszkiewicz/GoogleDrive/UW/GitHub/NextGenNEPA_LOCAL/Output/","cutadapt_output","/run",params$run.num,"_",format(Sys.Date(), "%Y%m%d"), "/noprimers")

```

# Split primer-trimmed fastq files from cutadapt by marker 

Some runs have 16S files as well. If the run doesn't, comment them out or it may just give you an error saying none were found.

```{r find all files and split by marker}

# use the pattern to sort out the files - ^ means "starts with" so we can use that to separate by marker
COI_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_COI", full.names = T)
MF_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_MiFish", full.names = T)
MM_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_MiMammal", full.names = T)
Ac16S_files <- list.files(path = fastq_noprimers_folder, pattern = "^Locus_Ac16S", full.names = T)

# then we actually want to create subfolders for each marker and put move the files into each folder by marker
new_COI_folder <- paste0(fastq_noprimers_folder,"/COI")
new_MF_folder <- paste0(fastq_noprimers_folder,"/MiFish")
new_MM_folder <- paste0(fastq_noprimers_folder,"/MiMammal")
new_Ac_folder <- paste0(fastq_noprimers_folder, "/Ac16S")

# then actually create them
dir.create(new_COI_folder)
dir.create(new_MF_folder)
dir.create(new_MM_folder)
dir.create(new_Ac_folder)

# then we  need to move the files (which we already have listed out)
file.move(COI_files, new_COI_folder)
file.move(MF_files, new_MF_folder)
file.move(MM_files, new_MM_folder)
file.move(Ac16S_files, new_Ac_folder)

```

# Split metadata files 

Now we should also split the sequencing metadata file by marker to use as input into dada2

```{r split metadata file by marker}
cutadapt_output_metadata_path <- paste0(fastq_noprimers_folder, "/output.metadata.csv")
cutadapt_output_metadata_all <- read.csv(cutadapt_output_metadata_path)

# split up based on marker
cutadapt_output_metadata_COI <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "COI",]
cutadapt_output_metadata_MF <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "MiFish",]
cutadapt_output_metadata_MM <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "MiMammal",]
cutadapt_output_metadata_Ac <- cutadapt_output_metadata_all[cutadapt_output_metadata_all$Locus == "Ac16S",]

# write them all to csv files 
write.csv(cutadapt_output_metadata_COI, paste0(new_COI_folder, "/cutadapt_output_metadata_COI.csv"), row.names=FALSE)
write.csv(cutadapt_output_metadata_MF, paste0(new_MF_folder, "/cutadapt_output_metadata_MiFish.csv"), row.names=FALSE)
write.csv(cutadapt_output_metadata_MM, paste0(new_MM_folder, "/cutadapt_output_metadata_MiMammal.csv"), row.names=FALSE)
write.csv(cutadapt_output_metadata_Ac, paste0(new_Ac_folder, "/cutadapt_output_metadata_Ac16S.csv"), row.names=FALSE)

```

Now we can use the dada2 scripts for each marker separately! 